code/modules/atmospherics/machinery/air_alarm/_air_alarm.dm

```diff
@@ -0,0 +1,529 @@
+#define AALARM_REPORT_TIMEOUT 100
+
+/obj/machinery/airalarm
+	name = "air alarm"
+	desc = "A machine that monitors atmosphere levels. Goes off if the area is dangerous."
+	icon = 'icons/obj/monitors.dmi'
+	icon_state = "alarmp"
+	idle_power_usage = BASE_MACHINE_IDLE_CONSUMPTION * 0.05
+	active_power_usage = BASE_MACHINE_ACTIVE_CONSUMPTION * 0.02
+	power_channel = AREA_USAGE_ENVIRON
+	req_access = list(ACCESS_ATMOSPHERICS)
+	max_integrity = 250
+	integrity_failure = 0.33
+	armor_type = /datum/armor/machinery_airalarm
+	resistance_flags = FIRE_PROOF
+
+	/// Current alert level, found in code/__DEFINES/atmospherics/atmos_machinery.dm
+	/// AIR_ALARM_ALERT_NONE, AIR_ALARM_ALERT_MINOR, AIR_ALARM_ALERT_SEVERE
+	var/danger_level = AIR_ALARM_ALERT_NONE
+
+	var/datum/air_alarm_mode/selected_mode
+	///A reference to the area we are in
+	var/area/my_area
+
+	var/locked = TRUE
+	var/aidisabled = 0
+	var/shorted = 0
+	var/buildstage = AIRALARM_BUILD_COMPLETE // 2 = complete, 1 = no wires,  0 = circuit gone
+
+	///Represents a signel source of atmos alarms, complains to all the listeners if one of our thresholds is violated
+	var/datum/alarm_handler/alarm_manager
+
+	var/static/list/atmos_connections = list(COMSIG_TURF_EXPOSE = PROC_REF(check_air_dangerlevel))
+
+	// An assoc list of [datum/tlv]s, indexed by "pressure", "temperature", and [datum/gas] typepaths.
+	var/list/datum/tlv/tlv_collection
+
+GLOBAL_LIST_EMPTY_TYPED(air_alarms, /obj/machinery/airalarm)
+
+/datum/armor/machinery_airalarm
+	energy = 100
+	fire = 90
+	acid = 30
+
+/obj/machinery/airalarm/Initialize(mapload, ndir, nbuild)
+	. = ..()
+	wires = new /datum/wires/airalarm(src)
+	if(ndir)
+		setDir(ndir)
+
+	if(nbuild)
+		buildstage = AIRALARM_BUILD_NO_CIRCUIT
+		set_panel_open(TRUE)
+
+	if(name == initial(name))
+		name = "[get_area_name(src)] Air Alarm"
+
+	tlv_collection = list()
+	tlv_collection["pressure"] = new /datum/tlv/pressure
```

Comment:
part of me is like "We could make these immutable!" but I bet it's like no cost/memory overhead, since they're just datums and rely on just init vars.
maybe better this way

Reply:
True. Not really worth the extra time coding too.